openpype/hosts/houdini/api/shelves.py

```diff
@@ -0,0 +1,188 @@
+import os
+import logging
+import platform
+
+from openpype.settings import get_project_settings
+
+import hou
+
+log = logging.getLogger("openpype.hosts.houdini")
+
+
+def generate_shelves():
+    """This function generates complete shelves from shef set to tools
+    in Houdini from openpype project settings houdini shelf definition.
+
+    Raises:
+        FileNotFoundError: Raised when the shelf set filepath does not exist
+    """
+    current_os = platform.system().lower()
+
+    # load configuration of houdini shelves
+    project_settings = get_project_settings(os.getenv("AVALON_PROJECT"))
+    shelves_set_config = project_settings["houdini"]["shelves"]
+
+    if not shelves_set_config:
+        log.warning(
+            "SHELF WARNGING: No custom shelves found in project settings."
+        )
+        return
+
+    for shelf_set_config in shelves_set_config:
+        shelf_set_filepath = shelf_set_config.get('shelf_set_source_path')
+
+        if shelf_set_filepath[current_os]:
+            if not os.path.isfile(shelf_set_filepath[current_os]):
+                raise FileNotFoundError(
+                    "SHELF ERROR: This path doesn't exist - {}".format(
+                        shelf_set_filepath[current_os]
+                    )
+                )
+
+            hou.shelves.newShelfSet(file_path=shelf_set_filepath[current_os])
+            continue
+
+        shelf_set_name = shelf_set_config.get('shelf_set_name')
+        if not shelf_set_name:
+            log.warning(
+                "SHELF WARNGING: No name found in shelf set definition."
+            )
+            return
+
+        shelf_set = get_or_create_shelf_set(shelf_set_name)
+
+        shelves_definition = shelf_set_config.get('shelf_definition')
+
+        if not shelves_definition:
+            log.warning(
+                "SHELF WARNING: \
+No shelf definition found for shelf set named '{}'".format(shelf_set_name)
+            )
+            return
+
+        for shelf_definition in shelves_definition:
+            shelf_name = shelf_definition.get('shelf_name')
+            if not shelf_name:
+                log.warning(
+                    "SHELF WARNGING: No name found in shelf definition."
+                )
+                return
+
+            shelf = get_or_create_shelf(shelf_name)
+
+            for tool_definition in shelf_definition.get('tools_list'):
+                # We verify that the name and script attibutes of the tool
+                # are set
+                mandatory_attributes = ['name', 'script']
```

Comment:
What is the point to define this as a set here?

Reply:
You have logic that does a "is value in list" check which is `On` complexity to run, by using a `set` instead that should be an `O1` complexity lookup and faster. (The list in this case was two values so speed difference might be negatable, but usually  you'd do lookups for values using a `set` because it uses the hash to check instead of iterating over the list until it finds the exact values).

E.g. [more info here](https://towardsdatascience.com/faster-lookups-in-python-1d7503e9cd38).

Reply:
Ok thank you, this is good to know !